### PR TITLE
fix(kda): pad fused in_proj past the hipBLASLt bf16 dead zone

### PR DESCRIPTION
**Pull request by vanshbhatia-amd**
## Summary

hipBLASLt on gfx950 has no bf16 GEMM solution for output widths in roughly `[2176, 2304]` once M (tokens) exceeds a single micro-batch. The 1B KDA config's fused `in_proj` width is `qk*2 + v + 2*head_v + num_heads = 2192`, landing squarely in that band, so the config raised `HIPBLASLT Error: 6` above `micro_batch_size: 2`. Empirically 2192 and 2304 fail while 2048, 2432 and 2560 succeed at every M.

This rounds widths **that fall inside the band** up to the next multiple of `kda_in_proj_pad_multiple` (default 512, so 2192 becomes 2560) and slices the extra columns off in `forward`. Checkpoint sharding keeps them as a named `pad` chunk so state dicts still round-trip. GDN is unaffected, since its projection widths are already powers of two.

## Why padding is gated on the band

Padding every width unconditionally is marginally faster. Measured as a same-node, same-session A/B on KDA-pure 300M, micro-batch 128, 500 steps:

| `in_proj` width | ms/iter | tok/s/GPU | peak VRAM | loss@500 |
|---|---|---|---|---|
| padded, 1160 -> 1536 | 701.2 | 373,850 | 205.6 GB | 5.181343 |
| native 1160 | 710.1 | 369,165 | **203.4 GB** | 5.181674 |

Padding is 1.3% quicker despite computing 32% more columns, because 1536 is a clean multiple of 512 while 1160 (8*145) tiles poorly.

That is not worth taking outside the band, because padding changes `in_proj`'s parameter shape. It costs 2.2 GB of peak activation memory, adds roughly 4.6M dead columns (376 x 1024 x 12 KDA layers) that carry optimizer state and are then discarded, and — most importantly — makes checkpoints trained without the padding fail to load. The released KDA 300M checkpoint (`zebra_llama_300M_kda_pure-10BT/iter_0004768`) stores `in_proj.weight` at (1160, 1024), so gating the padding keeps it loadable as-is.

Inside the band there is no such choice: the GEMM has no solution at all, so the shape change is unavoidable. No released checkpoint is affected there, since the only trained KDA model is the 300M one.

## Validation

- 1B KDA (width 2192 -> 2560) runs above `micro_batch_size: 2` with zero hipBLASLt errors.
- 300M KDA (width 1160) is left untouched by the gate, and the released checkpoint's `in_proj` shape still matches the built model.
- Both A/B arms above train to essentially the same step-500 loss (5.1813), the small difference being the expected numerics of a different GEMM shape.
- The two arms were also run on separate nodes; the cross-node ratio matched an independent GDN control to within 0.02 percentage points, so the 1.3% is a real effect rather than node drift.

## Export path

The pad columns are saved as a named `pad` chunk so state dicts round-trip unchanged. They carry no signal — `forward` slices them off, so their gradient is identically zero — and `convert_kda_to_fla_hf.py` now drops them when exporting to FLA rather than failing its exact-shape assert, which is what a padded 1B KDA checkpoint would previously have hit.

### CODE REVIEW

File: primus/backends/megatron/core/models/hybrid/kimi_delta_attention.py [hunk 1]
```diff
@@ -567,11 +604,13 @@ def forward(
         # the per-TP local dims; sum equals self.in_proj_dim // tp_size.
         qkv_local = self.qk_dim_local_tp * 2 + self.v_dim_local_tp
```
[the remark this answers is not in the document]
Comment (vanshbhatia-amd) [reply primus/backends/megatron/core/models/hybrid/kimi_delta_attention.py]: Good catch, fixed. The comment now states that the slice sizes sum to `self.in_proj_dim_padded // tp_size`, and that this is the same as `self.in_proj_dim // tp_size` whenever no padding was applied, so the padded case is no longer misdescribed.

File: primus/backends/megatron/core/models/hybrid/kimi_delta_attention.py [hunk 2]
```diff
@@ -308,10 +308,47 @@ def __init__(
...
+        # carry optimizer state and are then discarded. If memory is the binding
+        # constraint, set kda_in_proj_pad_multiple to 0 or 1 to keep the native
+        # width — but note that widths inside the band will then fail outright.
+        pad_multiple = getattr(self.config, "kda_in_proj_pad_multiple", 512)
+        if pad_multiple and pad_multiple > 1:
+            padded = ((self.in_proj_dim + pad_multiple - 1) // pad_multiple) * pad_multiple
+        else:
+            padded = self.in_proj_dim
+        self.in_proj_pad = padded - self.in_proj_dim
+        self.in_proj_dim_padded = padded
```
[the remark this answers is not in the document]
Comment (vanshbhatia-amd) [reply primus/backends/megatron/core/models/hybrid/kimi_delta_attention.py:357]: This was a real break, and the PR now avoids it.

I checked the released KDA checkpoint, `zebra_llama_300M_kda_pure-10BT/iter_0004768`: it stores `decoder.layers.N.mixer.in_proj.weight` at (1160, 1024) across all 12 KDA layers, so with unconditional 512-rounding the model would have built that projection at (1536, 1024) and the checkpoint would no longer have loaded. That is also the checkpoint behind the KDA downstream eval numbers, so it matters.

Rather than add a migration path, the padding is now gated to widths that actually land in the dead zone. The 300M width (1160) is left untouched and its checkpoint loads unchanged; the 1B width (2192) still pads to 2560, where the shape change is unavoidable because the GEMM has no solution at all, and no released checkpoint exists for that model. The only cost is the 1.3% that unconditional padding was buying at 300M, which is documented in the code and the PR description.

File: primus/backends/megatron/core/models/hybrid/kimi_delta_attention.py [hunk 3]
```diff
@@ -810,22 +857,29 @@ def sharded_state_dict(self, prefix="", sharded_offsets=(), metadata=None, tp_gr
...
+            self.v_dim // self.tp_size,
+            self.head_dim,  # f_a (bottleneck dim, not sharded)
+            self.head_dim,  # g_a (bottleneck dim, not sharded)
+            self.num_heads // self.tp_size,  # beta
+        ]
+        in_proj_split_names = ["query", "key", "value", "f_a", "g_a", "beta"]
+        if self.in_proj_pad:
+            # hipBLASLt dead-zone pad columns (see __init__); kept as a named
+            # chunk so the checkpoint round-trips, discarded at load for FLA export.
+            in_proj_split_sections.append(self.in_proj_pad // self.tp_size)
```
[the remark this answers is not in the document]
Comment (vanshbhatia-amd) [reply primus/backends/megatron/core/models/hybrid/kimi_delta_attention.py]: Correct on both counts, and the second one turned out to matter more than the comment did. There was no discard logic anywhere, and `convert_kda_to_fla_hf.py` asserts `in_proj_w.shape == (fused_in_proj_dim, hidden_size)` exactly, so a padded checkpoint would not merely round-trip with a stray chunk, it would fail to export at all. That is reachable for the 1B KDA config, which is the one this PR pads and unblocks.

So rather than only reword the comment, the converter now accepts a projection with trailing pad rows and slices them off before the split, keeping the shape check for genuinely wrong shapes. The comment now describes the real behavior: the chunk is saved so the checkpoint round-trips as-is, the columns carry no signal because forward slices them off and their gradient is identically zero, and the FLA export drops them.

File: primus/backends/megatron/core/models/hybrid/kimi_delta_attention.py [hunk 4]
```diff
@@ -810,22 +857,31 @@ def sharded_state_dict(self, prefix="", sharded_offsets=(), metadata=None, tp_gr
         # f_a/g_a have dim 64 (head_v_dim, NOT TP-sharded since head_v_dim < tp_size
         # in some configs — we still split per-rank as standard column-parallel),
         # beta has num_heads dim which is TP-sharded the same way.
-        in_proj_dim_local_tp = self.in_proj_dim // self.tp_size
+        in_proj_dim_local_tp = self.in_proj_dim_padded // self.tp_size
         assert sharded_state_dict[f"{prefix}in_proj.weight"].data.size(0) == in_proj_dim_local_tp, (
```
[the remark this answers is not in the document]
Comment (vanshbhatia-amd) [reply primus/backends/megatron/core/models/hybrid/kimi_delta_attention.py]: Fair point, that comment contradicted itself. `__init__` does hard-assert `tp_size == 1` for the fused `in_proj` (with a note on the TP>1 recipe), so the sharding narrative there was describing a case that cannot currently occur.

Reworded: it now says every section is its full width and the `// self.tp_size` divisions are no-ops today, kept only so the layout stays correct if that TP>1 recipe is ever implemented, at which point q/k/v and beta would shard along the output dim while `f_a`/`g_a` stay replicated as head_v_dim bottlenecks. That keeps the forward-looking intent without implying sharding happens now.

File: primus/backends/megatron/core/models/hybrid/kimi_delta_attention.py [hunk 5]
```diff
@@ -308,10 +312,49 @@ def __init__(
...
+        # parameter shape: that costs 2.2 GB of peak memory, adds ~4.6M dead
+        # columns carrying optimizer state, and makes checkpoints trained
+        # without the padding fail to load. Inside the band the GEMM has no
+        # solution at all, so there the shape change is unavoidable.
+        pad_multiple = getattr(self.config, "kda_in_proj_pad_multiple", 512)
+        lo, hi = _HIPBLASLT_BF16_DEAD_ZONE
+        if pad_multiple and pad_multiple > 1 and lo <= self.in_proj_dim <= hi:
+            padded = ((self.in_proj_dim + pad_multiple - 1) // pad_multiple) * pad_multiple
+        else:
+            padded = self.in_proj_dim
```
[the remark this answers is not in the document]
Comment (vanshbhatia-amd) [reply primus/backends/megatron/core/models/hybrid/kimi_delta_attention.py:355]: You are right that the trigger is narrower than the gate, but making the gate match it would trade a small waste for a real hazard, so I have documented the reasoning rather than changed the condition.

A parameter's shape has to be a function of the model config and nothing else. The same checkpoint is written in bf16 on gfx950 and then read back on CPU by `tools/hybrid/convert_kda_to_fla_hf.py` (it loads with `map_location="cpu"`), or in a different dtype for eval. If the width also depended on the runtime dtype or on a device-arch probe, `in_proj` would be 2560 at train time and 2192 at conversion time on the very same model, which breaks exactly the checkpoint compatibility your earlier comment was protecting — a worse failure than the one being avoided, and a nondeterministic one.

So the cost of the wider gate is bounded and visible: a config on another platform whose fused width happens to land in [2176, 2304] carries a few unused columns, and `kda_in_proj_pad_multiple: 0` opts out. The comment now states this explicitly so it reads as a deliberate choice rather than an oversight.
